src/sys/mips/syscall/mod.rs

```diff
@@ -22,31 +20,34 @@ pub(crate) use arch::{
 )]
 mod arch;
 
-pub(crate) use crate::sys::reg::{ParamRegR, ParamRegW, RetReg};
+pub use crate::sys::reg::{ParamRegR, ParamRegW, RetReg};
 
 /// Semihosting operation code.
-#[derive(Clone, Copy)]
-#[repr(usize)]
-#[non_exhaustive]
-pub(crate) enum OperationCode {
-    UHI_exit = 1,
-    UHI_open = 2,
-    UHI_close = 3,
-    UHI_read = 4,
-    UHI_write = 5,
-    UHI_lseek = 6,
-    UHI_unlink = 7,
-    UHI_fstat = 8,
-    UHI_argc = 9,
-    UHI_argnlen = 10,
-    UHI_argn = 11,
-    // UHI_ramrange = 12, // QEMU (as of 7.2) doesn't support this
-    // UHI_plog = 13, // TODO
-    // UHI_assert = 14, // TODO
-    // UHI_exception = 15, // QEMU (as of 7.2) doesn't support this
-    UHI_pread = 19,
-    UHI_pwrite = 20,
-    UHI_link = 22,
-    // UHI_boot_fail = 23, // QEMU (as of 7.2) doesn't support this
+#[derive(Debug, Clone, Copy)]
+pub struct OperationCode(usize);
+impl OperationCode {
+    pub const UHI_EXIT: Self = Self(1);
+    pub const UHI_OPEN: Self = Self(2);
+    pub const UHI_CLOSE: Self = Self(3);
+    pub const UHI_READ: Self = Self(4);
+    pub const UHI_WRITE: Self = Self(5);
+    pub const UHI_LSEEK: Self = Self(6);
+    pub const UHI_UNLINK: Self = Self(7);
+    pub const UHI_FSTAT: Self = Self(8);
+    pub const UHI_ARGC: Self = Self(9);
+    pub const UHI_ARGNLEN: Self = Self(10);
+    pub const UHI_ARGN: Self = Self(11);
+    // const UHI_RAMRANGE : Self = Self(12); // QEMU (as of 7.2) doesn't support this
+    // const UHI_PLOG : Self = Self(13); // TODO
+    // const UHI_ASSERT : Self = Self(14); // TODO
+    // const UHI_EXCEPTION : Self = Self(15); // QEMU (as of 7.2) doesn't support this
+    pub const UHI_PREAD: Self = Self(19);
+    pub const UHI_PWRITE: Self = Self(20);
+    pub const UHI_LINK: Self = Self(22);
+    // const UHI_BOOT_FAIL : Self = Self(23); // QEMU (as of 7.2) doesn't support this
+
+    #[inline] // inline to help compiler to remove assertion
+    pub const fn user_defined(number: usize) -> Self {
+        Self(number)
+    }
```

Comment:
```suggestion
```

It is not clear if UHI supports user-defined system calls, so I would prefer not to provide any for now.